### ISSUE TITLE
Files not overwritten when using development environment with manifest plugin

### ISSUE THREAD

**Issue by ip2k** [ember-cli-deploy/ember-cli-deploy-s3#48]
Hello,

I'm doing lightning deploys (index.html to Redis served by nginx with a redis module and everything else to S3) and seeing that with the `manifest` plugin enabled and `ENV.build.environment` in `config/deploy.js` set to `development`, since files are NOT fingerprinted and renamed, the manifest file matches everything (since the file names stay the same in development mode), and nothing gets uploaded. I can work around this by deleting everything in my target S3 location, or by manually enabling all of the plugins in the lightning pack except for `manifest`.

Here's my set-up:

```
$ ember -v --verbose
version: 2.3.0
http_parser: 2.6.0
node: 5.5.0
v8: 4.6.85.31
uv: 1.8.0
zlib: 1.2.8
ares: 1.10.1-DEV
icu: 56.1
modules: 47
openssl: 1.0.2e
os: darwin x64
```

```
# cat package.json
{
  "name": "ember-cli-skeleton",
  "version": "0.0.2",
  "description": "Small description for ember-cli-skeleton goes here",
  "private": true,
  "directories": {
    "doc": "doc",
    "test": "tests"
  },
  "scripts": {
    "build": "ember build",
    "start": "ember server",
    "test": "ember test"
  },
  "repository": "",
  "engines": {
    "node": ">= 0.10.0"
  },
  "author": "",
  "license": "MIT",
  "devDependencies": {
    "broccoli-asset-rev": "^2.2.0",
    "ember-ajax": "0.7.1",
    "ember-cli": "2.3.0",
    "ember-cli-app-version": "^1.0.0",
    "ember-cli-babel": "^5.1.5",
    "ember-cli-dependency-checker": "^1.2.0",
    "ember-cli-deploy": "0.5.1",
    "ember-cli-deploy-build": "0.1.1",
    "ember-cli-deploy-lightning-pack": "0.4.0",
    "ember-cli-deploy-slack": "0.1.0",
    "ember-cli-htmlbars": "^1.0.1",
    "ember-cli-htmlbars-inline-precompile": "^0.3.1",
    "ember-cli-inject-live-reload": "^1.3.1",
    "ember-cli-qunit": "^1.2.1",
    "ember-cli-release": "0.2.8",
    "ember-cli-sri": "^2.0.0",
    "ember-cli-uglify": "^1.2.0",
    "ember-data": "^2.3.0",
    "ember-disable-proxy-controllers": "^1.0.1",
    "ember-export-application-global": "^1.0.4",
    "ember-load-initializers": "^0.5.0",
    "ember-resolver": "^2.0.3",
    "loader.js": "^4.0.0"
  }
}
```

```
#cat config/deploy.js |egrep -v '(^|\s+)//.*$|/\*(.|.\s)*?\*/'
var VALID_DEPLOY_TARGETS = [
  'dev',
  'prod'
];

module.exports = function(deployTarget) {
  var ENV = {
    build: {},
    redis: {
      host: 'REDACTED',
      allowOverwrite: true,
      keyPrefix: 'ember-cli-skeleton:index'
    },
    s3: {
      accessKeyId: process.env.AWS_KEY,
      secretAccessKey: process.env.AWS_SECRET,
      bucket: 'REDACTED',
      region: 'us-west-2',
      prefix: 'ember-cli-skeleton'
    }
  };

  ENV["revision-data"] = {
    type: 'git-commit'
  };

  if (VALID_DEPLOY_TARGETS.indexOf(deployTarget) === -1) {
    throw new Error('Invalid deployTarget ' + deployTarget);
  }

  if (deployTarget === 'dev') {
    ENV.build.environment = 'development';
    ENV.redis.keyPrefix = 'ember-cli-skeleton-dev:index';
    ENV.s3.prefix = 'ember-cli-skeleton-dev';
    ENV.plugins = ['build', 'revision-data', 'redis', 's3'];
  }

  if (deployTarget === 'prod') {
    ENV.build.environment = 'production';
  }

  return ENV;
}
```

With the above config specifying `ENV.plugins = ['build', 'revision-data', 'redis', 's3'];` when `deployTarget === 'dev'`, things work as expected and it skips the manifest step, so ALL files are re-uploaded to S3 and the site will update when I make changes and run `ember deploy dev --activate`. Commenting that line and letting all plugins run brings back the bug.

Please let me know if it'd be better to open this against the manifest plugin. I'm not entirely sure what the correct behavior in this case would be, but a plugin option to re-upload everything (overwriting duplicate files) or ignore the manifest are my naive ideas.

Thanks!

**Comment by achambers** [comment]
Hi @ip2k, I don't really see this as a bug. The manifest plugin is doing exactly what it's intended to do, not upload files that already exist. What you really want is to not run the manifest plugin when you're deploying to dev, which is what you described above with the `plugins` array.

Unfortunately the `plugins` array is the only way to current restrict a plugin running for a particular deploy target. However we are working on a much more powerful way to enable and disable plugins based on some criteria which means you would essentially be able to do something along the lines of set the manifest plugin to be disabled for `dev`. But this capability does not yet exist, which leaves the `plugin` array.

So, tl;dr; here is that I don't really see this as a bug at all. Thoughts?

**Comment by lukemelia** [comment]
Generally, I agree with @achambers. Also wanted to mention that in our case of using lightning pack, we don't do S3 uploads in development, we serve code locally from ember-cli.

**Comment by achambers** [comment]
> we don't do S3 uploads in development, we serve code locally from ember-cli

As do I

**Comment by achambers** [comment]
I am going to close this issue based on the reasoning above. If you have anything you'd like to clarify @ip2k, or you think this should be reopened, just holler.

**Comment by ip2k** [comment]
Thanks all for the fast responses. I look forward to a better way to enable / disable plugins :)
